src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyDiscoveryManager.cs

```diff
@@ -104,6 +153,37 @@ public void Initialize(bool skipDefaultAdapters)
         /// <inheritdoc/>
         public void DiscoverTests(DiscoveryCriteria discoveryCriteria, ITestDiscoveryEventsHandler2 eventHandler)
         {
+            if (this.proxyOperationManager == null)
+            {
+                // In case we have an active test session, we always prefer the already
+                // created proxies instead of the ones that need to be created on the spot.
+                this.proxyOperationManager = TestSessionPool.Instance.TakeProxy(
```

Comment:
I find having a static `TestSessionPool` instance to be the easiest way of getting access to sessions and I see no disadvantages to this approach for now. All sessions corresponding to a single instance of `vstest.console` can be managed by the same `TestSessionPool` so there's no need for multiple instances of `TestSessionPool`.

The disadvantages of removing the `static` qualifier though are quite large and the benefits are not apparent. For example, we'll have to do a lot of wiring to manage the `TestSessionPool` at the `DesignModeClient` level probably and then pass it down through all TP layers to where it is actually needed when creating the test session. Then, we'll have to do the same thing for test discovery/run, because those operations will have to get an already running testhost in order to get the performance benefits we discussed. All this will make the code harder to maintain in my opinion by passing down yet another argument, whereas thinking of the `TestSessionPool` as having the same lifetime as the `vstest.console` would add no extra maintainability challenges.